### DISCUSSION TITLE
Another idea for memory layout

### DISCUSSION THREAD

**Discussion by DL8** [amichai-bd/riscv-multi-core-lotr#49] [Ideas]
I thought of (yet another) possible memory layout. This one will allow access to shared memory and CR space. It has the following basic structure:

| ***TYPE*** | ***ADDRESS*** |
| --- | --- |
| 31 | 30:0 |

Where `TYPE` indicates whether this is global or agent-specific space.

For agent memory, the layout will be similar to the current:
| ***0*** | ***AGENT_ID*** | ***REGION*** | ***OFFSET*** |
| --- | --- | --- | --- |
| 31 | 30:23 | 22:21 | 20:0 |

For shared memory, it will look like that:
| ***1*** | ***REGION*** | ***OFFSET*** |
| --- | --- | --- |
| 31 | 30 | 29:0 |

Where `REGION` is either 0 (memory) or 1 (CR space).

This design allows having a global CR space and potentially bigger GPC shared memory. Also, it makes sense to have a global CR space for global GPC configurations (e.g. number of agent ID bits). I believe shared memory may also be DRAM, so the main CPU can share data with GPC (but I'm not sure how complicated it is or if it's really needed. Also, I'm not sure if it should be in addition to or instead of GPC shared memory).

**Comment by amichai-bd** [comment]
I will take it a step further.
For the "Shared memory" (bit 31=1'b1) - when you look only at bit 30 to determine CR Space, you are allocating 1 GB of memory just for CRs.
So in the case of shared memory (address[31]=1'b1), we should access the "shared memory".
Threre - we loot at bits Address[30:12] .  if its 19'h7ffff will indicate the CR Space. (4KB of "global CR Space" [11:0].

But yet again - I don't think we need anything at that scale for our POC.
I think using the Agent ID bits [31:24] will allow us to have 256 different "agents" - Some Agents may be just a Shared non-local memory. effectively giving each agent up to 8MB of memory.
And in case we want a bigger memory the same Agent may get more "ID" mapping to the same Agent. (we can simply "mask" out LSB Agent ID bits so multiple ID addresses will match the same Agent.
